src/gc.c

```diff
@@ -1316,14 +1324,23 @@ STATIC_INLINE jl_value_t *jl_gc_pool_alloc_inner(jl_ptls_t ptls, int pool_offset
     return jl_valueof(v);
 }
 
-// Instrumented version of jl_gc_pool_alloc_inner, called into by LLVM-generated code.
+// Deprecated version, supported for legacy code.
 JL_DLLEXPORT jl_value_t *jl_gc_pool_alloc(jl_ptls_t ptls, int pool_offset,
                                           int osize)
 {
     jl_value_t *val = jl_gc_pool_alloc_inner(ptls, pool_offset, osize);
     maybe_record_alloc_to_profile(val, osize, jl_gc_unknown_type_tag);
     return val;
 }
+// Instrumented version of jl_gc_pool_alloc_inner, called into by LLVM-generated code.
+JL_DLLEXPORT jl_value_t *jl_gc_pool_alloc_typed(jl_ptls_t ptls, int pool_offset,
+                                        int osize, jl_value_t* type)
+{
+    jl_value_t *val = jl_gc_pool_alloc_inner(ptls, pool_offset, osize);
+    jl_set_typeof(val, type);
+    maybe_record_alloc_to_profile(val, osize, (jl_datatype_t*)type);
```

Comment:
The other variant on this PR could be that we _only_ set record the alloc here, but don't actually call set_typeof, and keep the tag setting back in the generated code.

......... Do you think that would be a better change? It would keep things closer to how they were before this, and i don't see how it could be any _worse_, performance-wise. And it might be possible to do better than this generic C function if you know the context?

Reply:
On the other hand, it means compiling that instruction over and over, which means longer compile times. And if it actually makes no difference, then factoring it out into this C function would make compilation slightly cheaper.

Reply:
K, i pushed that up as another variant for us to consider, in 45f3f5ea9b.

Reply:
I think I liked the previous version better, the reasoning being that it's kina weird to write at a negative offer of an object. Iirc we promise a certain amount of valid bytes and then we suddenly write to the bytes in front of it.

Reply:
I think I liked the change better too, but I like the idea of keeping the changes minimal I guess.

We're talking about it with @gbaraldi and @jpsamaroo in the Profiling call now - and Gabriel points out that there is potentially more optimization opportunity by exposing more to LLVM.

For example, if someone tries to read the type afterwards, can it elide the read? ... Though maybe that isn't even possible anyway since this would be volatile memory...

Reply:
We are already load and store the type tag a bunch from JITted and maybe this allows for an optimization.

Reply:
So the optimization that could happen is store-load forwarding. Iff LLVM can prove that no-one else could modify the value pointed to in-between (I do not see how it could prove that).

We already do tag forwarding.

Reply:
yeah, that's my feeling too Valentin. I don't know how it could prove that.

But i do like the idea of keeping this change more limited in scope.. :/

Reply:
But maybe that's the coward's answer, and I should Be Bold and make the change!? Dunno.

It does seem a bit wrong to be changing a fundamental part of codegen as part of a profiling PR though.....

Maybe we can start with this way, and then follow up with another PR that moves the store into the C function?

Reply:
You are already modifying codegen.

But maybe @keno or @vtjnash know why we do the store in emitted IR.

Reply:
Thanks for the review, @vtjnash. Can you chime in on the above question? Do you know why we do the store in the emitted IR, rather than in C? The existing comments imply that this was intentional, and it'd be great to know.

(I'll probably merge the PR this way, after adjusting the docs, and then open a separate PR to move it into C instead?)

Reply:
I think we used to use that for some type-tag load-store forwarding, but we do that differently now